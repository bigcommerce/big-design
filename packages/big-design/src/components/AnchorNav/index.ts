export { AnchorNav, type AnchorNavProps, type AnchorNavElement } from './AnchorNav';
