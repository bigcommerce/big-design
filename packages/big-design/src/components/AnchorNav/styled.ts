import { theme as defaultTheme } from '@bigcommerce/big-design-theme';
import styled from 'styled-components';

import { BoxProps } from '../Box';

export const StyledAnchorNav = styled.div<BoxProps>`
display: block;
position relative;

.anchor-nav {
    &__list {
        position: sticky;
        inset-block-start: 0;
        width: 266px;
        float: inline-start;
        display:none;

        @media (min-width: ${({ theme }) => theme.breakpointValues.tablet}) {
            display: block;
            margin-inline-end: ${defaultTheme.spacing.xLarge};
        }

        & > ul {
            list-style-type: none;
            padding: 0;
            margin: 0;

            & > li {
                display:block;
                margin: 0;
                padding: 0;
                box-sizing: border-box;

                & > a {
                    color: ${defaultTheme.colors.primary};
                    text-decoration: none;
                    display: block;
                    padding: ${defaultTheme.spacing.small} ${defaultTheme.spacing.large};
                    border-inline-start: 3px solid transparent;

                    &:hover {
                        background-color: ${defaultTheme.colors.primary10};
                    }

                    &.active {
                        color: ${defaultTheme.colors.secondary70};
                        border-inline-start: 3px solid ${defaultTheme.colors.primary};

                        &:hover {
                            background-color: transparent;
                        }
                    }
                }
            }
        }
    
    }
    
    &__elements {
        display:flex;
        flex-direction: column;
        gap: ${defaultTheme.spacing.xLarge};
    }
}
`;

StyledAnchorNav.defaultProps = { theme: defaultTheme };
