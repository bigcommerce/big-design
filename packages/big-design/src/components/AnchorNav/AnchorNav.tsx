import React, { useCallback, useEffect, useRef, useState } from 'react';
import { useInView } from 'react-intersection-observer';

import { Box } from '../Box';
import { Flex, FlexItem } from '../Flex';

import { StyledAnchorNavList } from './styled';

export interface AnchorNavElement {
  id: string;
  label: string;
  element: React.ReactNode;
}

export interface AnchorNavProps {
  readonly elements: AnchorNavElement[];
}

const useAnchorObserver = (
  id: string,
  setActiveId: React.Dispatch<React.SetStateAction<string | null>>,
  suspendObserverRef: React.MutableRefObject<boolean>,
): React.RefCallback<HTMLElement> => {
  const { ref, inView } = useInView({
    threshold: 0.1,
    rootMargin: '0px 0px -90% 0px',
  });

  useEffect(() => {
    if (inView && !suspendObserverRef.current) {
      setActiveId(id);
    }
  }, [inView, id, setActiveId, suspendObserverRef]);

  return ref;
};

interface AnchorSectionProps {
  readonly id: string;
  readonly children: React.ReactNode;
  readonly setActiveId: React.Dispatch<React.SetStateAction<string | null>>;
  readonly suspendObserverRef: React.MutableRefObject<boolean>;
  readonly registerSection: (id: string, el: HTMLElement | null) => void;
}

const AnchorSection: React.FC<AnchorSectionProps> = ({
  id,
  children,
  setActiveId,
  suspendObserverRef,
  registerSection,
}) => {
  // We now know that observerRef is a function callback.
  const observerRef = useAnchorObserver(id, setActiveId, suspendObserverRef);

  // Combine the observer ref with the registration of the section.
  const combinedRef = useCallback(
    (el: HTMLElement | null) => {
      registerSection(id, el);
      observerRef(el); // Always call the ref function directly.
    },
    [id, registerSection, observerRef],
  );

  return (
    <FlexItem id={id} ref={combinedRef} style={{ scrollMarginTop: '80px' }}>
      {children}
    </FlexItem>
  );
};

export const AnchorNav: React.FC<AnchorNavProps> = ({ elements }) => {
  const [activeId, setActiveId] = useState<string | null>(null);
  const sectionRefs = useRef<Record<string, HTMLElement | null>>({});
  const suspendObserverRef = useRef(false);
  const resumeTimeoutRef = useRef<ReturnType<typeof setTimeout> | null>(null);

  const updateUrlHash = useCallback((id: string) => {
    const newHash = `#${id}`;

    if (window.location.hash !== newHash) {
      history.replaceState(null, '', newHash); // Avoids adding to browser history
    }
  }, []);

  const scrollToSection = useCallback(
    (id: string) => {
      const el = sectionRefs.current[id];

      if (el) {
        el.scrollIntoView({ behavior: 'smooth' });
        setActiveId(id);
        updateUrlHash(id);
        suspendObserverRef.current = true;

        if (resumeTimeoutRef.current) {
          clearTimeout(resumeTimeoutRef.current);
        }

        resumeTimeoutRef.current = setTimeout(() => {
          suspendObserverRef.current = false;
        }, 1000);
      }
    },
    [updateUrlHash],
  );

  // Scroll to section based on the URL hash when the component loads.
  useEffect(() => {
    const hash = window.location.hash?.slice(1);

    if (hash && sectionRefs.current[hash]) {
      scrollToSection(hash);
    }
  }, [scrollToSection]);

  // Cleanup timeout on unmount.
  useEffect(() => {
    return () => {
      if (resumeTimeoutRef.current) {
        clearTimeout(resumeTimeoutRef.current);
      }
    };
  }, []);

  return (
    <Box>
      <StyledAnchorNavList>
        <ul>
          {elements.map(({ id, label }) => (
            <li key={id}>
              <a
                aria-current={id === activeId ? 'true' : undefined}
                className={id === activeId ? 'active' : ''}
                href={`#${id}`}
                onClick={(e) => {
                  e.preventDefault();
                  scrollToSection(id);
                }}
              >
                {label}
              </a>
            </li>
          ))}
        </ul>
      </StyledAnchorNavList>

      <Flex flexDirection="column" flexGap="xLarge">
        {elements.map(({ id, element }) => (
          <AnchorSection
            id={id}
            key={id}
            registerSection={(id, el) => {
              sectionRefs.current[id] = el;
            }}
            setActiveId={setActiveId}
            suspendObserverRef={suspendObserverRef}
          >
            {element}
          </AnchorSection>
        ))}
      </Flex>
    </Box>
  );
};

AnchorNav.displayName = 'AnchorNav';
