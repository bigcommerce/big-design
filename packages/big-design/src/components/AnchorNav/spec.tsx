import { render } from '@testing-library/react';
import React from 'react';
import 'jest-styled-components';

import { AnchorNav, AnchorNavElement } from './AnchorNav';

const intersectionObserverMock = () => ({
  observe: () => null,
  unobserve: () => null,
  disconnect: () => null,
});

window.IntersectionObserver = jest.fn().mockImplementation(intersectionObserverMock);
window.HTMLElement.prototype.scrollIntoView = jest.fn();

beforeEach(() => {
  // Reset the hash before each test to avoid side effects
  window.location.hash = '';
});

test('renders default AnchorNav', () => {
  const elements: AnchorNavElement[] = [
    { id: 'section1', label: 'Section 1', element: <div>Content 1</div> },
    { id: 'section2', label: 'Section 2', element: <div>Content 2</div> },
  ];

  const { container } = render(<AnchorNav elements={elements} />);

  expect(container.firstChild).toMatchSnapshot();
  expect(container.querySelector('.anchor-nav__list')).toBeInTheDocument();
});

test('renders AnchorNav with active section', () => {
  const elements: AnchorNavElement[] = [
    { id: 'section1', label: 'Section 1', element: <div>Content 1</div> },
    { id: 'section2', label: 'Section 2', element: <div>Content 2</div> },
  ];

  // Simulate an active section by setting the hash
  window.location.hash = '#section2';

  const { container } = render(<AnchorNav elements={elements} />);

  expect(container.querySelector('.anchor-nav__list')).toHaveTextContent('Section 2');
});

test('should handle section refs correctly', () => {
  const elements: AnchorNavElement[] = [
    { id: 'section1', label: 'Section 1', element: <div>Content 1</div> },
    { id: 'section2', label: 'Section 2', element: <div>Content 2</div> },
  ];

  const { container } = render(<AnchorNav elements={elements} />);

  expect(container.querySelector('.anchor-nav__list')).toBeInTheDocument();

  // Check if the section refs are being set correctly
  const section1 = document.getElementById('section1');
  const section2 = document.getElementById('section2');

  expect(section1).toBeInTheDocument();
  expect(section2).toBeInTheDocument();
});
