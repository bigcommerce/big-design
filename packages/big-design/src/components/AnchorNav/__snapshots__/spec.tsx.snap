// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`AnchorNav Component matches the snapshot 1`] = `
.c0 {
  position: -webkit-sticky;
  position: sticky;
  -webkit-inset-block-start: 0;
  -ms-intb-rlock-start: 0;
  inset-block-start: 0;
  width: 16.625rem;
  float: inline-start;
  display: none;
}

.c0 > ul {
  list-style-type: none;
  padding: 0;
  margin: 0;
}

.c0 > ul > li {
  display: block;
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

.c0 > ul > li > a {
  color: #3C64F4;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: block;
  padding: 0.75rem 1.25rem;
  border-inline-start: 0.25rem solid transparent;
}

.c0 > ul > li > a:hover {
  background-color: #F0F3FF;
}

.c0 > ul > li > a.active {
  color: #313440;
  border-inline-start: 0.25rem solid #3C64F4;
}

.c0 > ul > li > a.active:hover {
  background-color: transparent;
}

@media (min-width:720px) {
  .c0 {
    display: block;
    margin-inline-end: 1.5rem;
  }
}

<nav
  class="c0"
>
  <ul>
    <li>
      <a
        class=""
        href="#section1"
      >
        Section 1
      </a>
    </li>
    <li>
      <a
        class=""
        href="#section2"
      >
        Section 2
      </a>
    </li>
  </ul>
</nav>
`;
